### PR TITLE
Clean up CustomForm, extract form types

### PR DESCRIPTION
**Pull request by kulyk**
Related to #22372: replacing `redux-form`

This PR extracts some common form types to `metabase-types/forms` and cleans up the `CustomForm` and `StandardForm` components. `CustomForm` and `StandardForm` are the most common way to render forms in Metabase.

`CustomForm` used to be a single JS file defining a few shared form components (`CustomFormField`, `CustomFormSubmit`, `CustomFormFooter`, etc.). This PR breaks it down into a few files and converts it into TypeScript without any API changes. Also, `CustomForm` uses legacy React context API to pass down some props. This PR isolates the components from legacy context usages to make it easier to eventually remove them.

### CODE REVIEW

**Comment by kulyk** [comment]
@metabase-bot run visual tests

File: frontend/src/metabase/components/form/CustomForm/types.ts [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+import {
+  FieldName,
+  DefaultFieldValue,
+  FieldValues,
+  FormFieldDefinition,
+  FormField,
+  FormObject,
+} from "metabase-types/forms";
+
+export interface BaseFormProps {
```
Comment (WiNloSt) [inline frontend/src/metabase/components/form/CustomForm/types.ts:11]: As far as I understand, some if not most of these are part of `redux-form` right? It's quite confusing where these might be coming from. Maybe adding a comment or using types directly from `@types/redux-form` would make it more obvious? What do you thinK?

**Review by WiNloSt — COMMENTED** [review]
I haven't fully reviewed the PR, but some of the types are a bit magical to me. Would it be possible to make those more obvious? e.g. some that are from `redux-form` and especially when it's from an older version of `redux-form`.

File: frontend/src/metabase/components/form/CustomForm/types.ts (hunk 1 above)
Comment (kulyk) [reply frontend/src/metabase/components/form/CustomForm/types.ts:11]: @WiNloSt yeah, form types are pretty heavy as there are a few different ways to render a form using the `Form` container:

**Entity form ([example](https://github.com/metabase/metabase/blob/17aa073720964f210bd84406a73a5f7422783722/frontend/src/metabase/collections/containers/CollectionCreate.jsx#L45))**

```jsx
<Collection.Form collection={collection} />
```

**Field definitions ([example](https://github.com/metabase/metabase/blob/17aa073720964f210bd84406a73a5f7422783722/frontend/src/metabase/query_builder/components/DatasetEditor/DatasetFieldMetadataSidebar/DatasetFieldMetadataSidebar.jsx#L296))**

```jsx
<Form
  form={{
    fields: [ { name: "name" } ]
  }}
/>
```

**Inline fields ([example](https://github.com/metabase/metabase/blob/17aa073720964f210bd84406a73a5f7422783722/enterprise/frontend/src/metabase-enterprise/auth/components/SettingsSAMLForm.jsx#L48))**

```jsx
<Form>
  <FormField name="name">
  <FormField name="description" type="text">
  <FormField name="collection" type="collection_id">
<Form/>
```

As a result, Form's props are pretty bloated as they handle a few different APIs that can also be used simultaneously. . A lot of props are coming from `redux-form`, I wouldn't actually highlight them as we're going to remove it. I'm trying to replace it without changing the external API too much, so we don't need to patch every single form in Metabase. So the new Form component is going to follow the same props structure / API and so the types will stay like this for some time

**Review by npfitz — APPROVED** [review]

**Review by daltojohnso — APPROVED** [review]

**Pull request merged**